### PR TITLE
Logrus for keysync

### PR DESCRIPTION
**Pull request by mcpherrinm**
There's still one small use of the stdlib logger, which we can replace after https://github.com/square/go-sq-metrics/pull/7 lands

There's some more work that can be done to make logs more structured, especially in client.go (which is code imported from keywhiz-fs).

### CODE REVIEW

File: api.go [hunk 1]
```diff
@@ -16,25 +16,25 @@ package keysync
 
 import (
 	"fmt"
-	"log"
 	"net/http"
-
 	"net/http/pprof"
+	"time"
 
-	"github.com/getsentry/raven-go"
```
Comment (worldwise001) [inline api.go]: not using sentry anymore?

File: api.go [hunk 2]
```diff
@@ -44,28 +44,32 @@ func (a *APIServer) syncAll(w http.ResponseWriter, r *http.Request) {
 func (a *APIServer) syncOne(w http.ResponseWriter, r *http.Request) {
 	client, hasClient := mux.Vars(r)["client"]
 	if !hasClient || client == "" {
+		a.logger.Info("Invalid request: No client provided.")
 		http.Error(w, "Invalid request: Please provide a client", http.StatusBadRequest)
```
Comment (worldwise001) [inline api.go]: Why are these messages different?

File: api.go [hunk 3]
```diff
@@ -16,25 +16,25 @@ package keysync
...
 type APIServer struct {
 	syncer *Syncer
+	logger *logrus.Entry
 }
 
 func (a *APIServer) syncAll(w http.ResponseWriter, r *http.Request) {
 	err := a.syncer.RunOnce()
 	if err != nil {
+		a.logger.WithError(err).Warn("Running syncer")
 		http.Error(w, fmt.Sprintf("Error syncing: %v", err), http.StatusInternalServerError)
```
Comment (worldwise001) [inline api.go]: Why are these messages different?

File: client.go [hunk 4]
```diff
@@ -110,43 +110,48 @@ func (c *Client) RebuildClient() error {
...
 // ServerStatus returns raw JSON from the server's _status endpoint
 func (c Client) ServerStatus() (data []byte, err error) {
+	logger := c.logger.WithField("logger", "_status")
 	now := time.Now()
 	t := *c.url
 	t.Path = path.Join(c.url.Path, "_status")
 	resp, err := c.httpClient.Get(t.String())
 	if err != nil {
-		c.Errorf("Error retrieving server status: %v", err)
+		logger.WithError(err).Warn("Retrieving server status")
```
Comment (worldwise001) [inline client.go]: These messages are different?

File: client.go [hunk 5]
```diff
@@ -110,43 +110,48 @@ func (c *Client) RebuildClient() error {
...
+	logger.WithFields(logrus.Fields{
+		"StatusCode": resp.StatusCode,
+		"duration":   time.Since(now),
+	}).Info("GET /_status")
 	defer resp.Body.Close()
 
 	data, err = ioutil.ReadAll(resp.Body)
 	if err != nil {
-		c.Errorf("Error reading response body for server status %v", err)
+		logger.WithError(err).Warn("Reading server status response")
```
Comment (worldwise001) [inline client.go]: These messages are also different?

File: cmd/keysync.go [hunk 6]
```diff
@@ -30,41 +30,65 @@ import (
...
 
 	raven.CapturePanicAndWait(func() {
-		metricsHandle := sqmetrics.NewMetrics(config.MetricsURL, config.MetricsPrefix, http.DefaultClient, 30*time.Second, metrics.DefaultRegistry, &log.Logger{})
+		metricsHandle := sqmetrics.NewMetrics(config.MetricsURL, config.MetricsPrefix, http.DefaultClient, 30*time.Second, metrics.DefaultRegistry, &stdlog.Logger{})
 
-		syncer, err := keysync.NewSyncer(config, metricsHandle)
+		syncer, err := keysync.NewSyncer(config, logger, metricsHandle)
 		if err != nil {
-			raven.CaptureErrorAndWait(err, nil)
+			logger.WithError(err).Fatal("Creating syncer")
```
Comment (worldwise001) [inline cmd/keysync.go]: If I'm about to fatal log, "Creating syncer" is not an informative comment.

File: cmd/keysync.go [hunk 7]
```diff
@@ -30,41 +30,65 @@ import (
...
 		// Start the API server
 		if config.APIPort != 0 {
-			keysync.NewAPIServer(syncer, config.APIPort)
+			keysync.NewAPIServer(syncer, config.APIPort, logger)
 		}
 
 		err = syncer.Run()
 		if err != nil {
-			raven.CaptureErrorAndWait(err, nil)
+			logger.WithError(err).Fatal("Running syncer")
```
Comment (worldwise001) [inline cmd/keysync.go]: Same here.

File: ownership.go [hunk 8]
```diff
@@ -33,46 +33,50 @@ type Ownership struct {
...
-		log.Printf("Error resolving uid for %v: %v\n", username, err)
-		return uint32(os.Geteuid())
+		return 0, fmt.Errorf("Error resolving uid for %v: %v\n", username, err)
 	}
 
 	uid, err := strconv.ParseUint(u.Uid, 10 /* base */, 32 /* bits */)
 	if err != nil {
-		log.Printf("Error resolving uid for %v: %v\n", username, err)
-		return uint32(os.Geteuid())
+		return 0, fmt.Errorf("Error resolving uid for %v: %v\n", username, err)
```
Comment (worldwise001) [inline ownership.go]: Do we want to make this a slightly different error message than above? It looks like this is just due to failing to parse as opposed to doing a lookup.

File: secret.go [hunk 9]
```diff
@@ -58,30 +57,35 @@ type Secret struct {
...
+func (s Secret) ModeValue() (os.FileMode, error) {
 	mode := s.Mode
 	if mode == "" {
 		mode = "0440"
 	}
 	modeValue, err := strconv.ParseUint(mode, 8 /* base */, 16 /* bits */)
 	if err != nil {
-		log.Printf("Unable to convert secret mode (%v) to octal, using '0440': %v\n", mode, err)
-		modeValue = 0440
+		return 0, fmt.Errorf("Unable to parse secret mode (%v): %v\n", mode, err)
```
Comment (worldwise001) [inline secret.go]: I would maybe add the word "file" somewhere in here, to note that it is a file mode.

File: syncer.go [hunk 10]
```diff
@@ -65,6 +65,8 @@ func (s *Syncer) LoadClients() error {
 	if err != nil {
 		return err
 	}
+	s.logger.WithField("count", len(newConfigs)).Info("Loaded configs")
```
Comment (worldwise001) [inline syncer.go]: What's the count field? Does this just show how many config lines were parsed? If so, can the comment reflect that?

File: syncer.go [hunk 11]
```diff
@@ -78,33 +80,28 @@ func (s *Syncer) LoadClients() error {
 		// Otherwise we (re)create the client
 		client, err := s.buildClient(name, clientConfig, s.metricsHandle)
 		if err != nil {
-			raven.CaptureError(err, nil)
-			fmt.Printf("Error building client %s: %v\n", name, err)
+			s.logger.WithError(err).WithField("client", name).Error("Building client")
```
Comment (worldwise001) [inline syncer.go]: These logging messages don't match.

File: syncer.go [hunk 12]
```diff
@@ -78,33 +80,28 @@ func (s *Syncer) LoadClients() error {
...
-	klogConfig := klog.Config{
-		Debug:      s.config.Debug,
-		Syslog:     false,
-		Mountpoint: name,
-	}
-	client, err := NewClient(clientConfig.Cert, clientConfig.Key, s.config.CaFile, s.server, time.Minute, klogConfig, metricsHandle)
+	clientLogger := s.logger.WithField("client", name)
+	client, err := NewClient(clientConfig.Cert, clientConfig.Key, s.config.CaFile, s.server, time.Minute, clientLogger, metricsHandle)
 	if err != nil {
-		fmt.Println("Error from NewClient", err)
```
Comment (worldwise001) [inline syncer.go]: Did we want to log this?

File: syncer.go [hunk 13]
```diff
@@ -115,7 +112,12 @@ func (s *Syncer) buildClient(name string, clientConfig ClientConfig, metricsHand
 	if group == "" {
 		group = s.config.DefaultGroup
 	}
-	writeConfig := WriteConfig{EnforceFilesystem: s.config.FsType, ChownFiles: s.config.ChownFiles, DefaultOwnership: NewOwnership(user, group)}
+	defaultOwnership, err := NewOwnership(user, group)
+	if err != nil {
+		// We log an error here but continue on.  The default of "0", root, is safe.
+		s.logger.WithError(err).Error("Default ownership")
```
Comment (worldwise001) [inline syncer.go]: This is not a descriptive message. What ended up getting default ownership?

File: syncer.go [hunk 14]
```diff
@@ -137,16 +139,16 @@ func (s *Syncer) Run() error {
 	for {
 		err = s.RunOnce()
 		if err != nil {
-			fmt.Printf("Error running sync: %+v\n", err)
-			raven.CaptureErrorAndWait(err, nil)
+			s.logger.WithError(err).Error("Running sync")
```
Comment (worldwise001) [inline syncer.go]: These logging messages don't match.

File: syncer.go [hunk 15]
```diff
@@ -162,7 +164,7 @@ func (s *Syncer) RunOnce() error {
 		err = entry.Sync()
 		if err != nil {
 			// Record error but continue updating other clients
-			raven.CaptureError(err, map[string]string{"name": name})
+			s.logger.WithError(err).WithField("name", name).Error("Syncing")
```
Comment (worldwise001) [inline syncer.go]: More descriptive error message?

File: syncer.go [hunk 16]
```diff
@@ -193,22 +195,21 @@ func (entry *syncerEntry) Sync() error {
 		name := filepath.Join(entry.Mountpoint, filename)
 		err = atomicWrite(name, secret, entry.WriteConfig)
 		if err != nil {
-			fmt.Printf("Couldn't write secret %s: %+v\n", secret.Name, err)
+			entry.logger.WithError(err).WithField("file", secret.Name).Error("Writing secret")
```
Comment (worldwise001) [inline syncer.go]: These logging messages don't match.

File: syncer.go (hunk 12 above)
Comment (mcpherrinm) [inline syncer.go]: No, it's logged in the caller.

File: syncer.go (hunk 10 above)
Comment (mcpherrinm) [inline syncer.go]: No, it shows the number of loaded configs

File: api.go (hunk 1 above)
Comment (mcpherrinm) [inline api.go]: not directly.

**Pull request merged**